Lib/python/pyerrors.swg

```diff
@@ -95,8 +95,12 @@ SWIG_Python_RaiseOrModifyTypeError(const char *message)
 #else
     newvalue = PyString_FromFormat("%s\nAdditional information:\n%s", PyString_AsString(value), message);
 #endif
-    Py_XDECREF(value);
-    PyErr_Restore(type, newvalue, traceback);
+    if (newvalue) {
+      Py_XDECREF(value);
+      PyErr_Restore(type, newvalue, traceback);
+    } else {
+      PyErr_Restore(type, value, traceback);
+    }
```

Comment:
I don't see any information about this change and why it has been done.

Reply:
I was reviewing the reference counting and spotted this as an out-of-memory \(OOM\) problem.

As described in [the Python/C API documentation for exception handling](https://docs.python.org/3/c-api/exceptions.html?highlight=pyerr_restore#c.PyErr_Fetch), `PyErr_Fetch()` and `PyErr_Restore()` operate as a pair, with `PyErr_Fetch()` acquiring a reference to `value` and `traceback` for the caller \(here `SWIG_Python_RaiseOrModifyTypeError()`\), and `PyErr_Restore()` handing references back. These then allow an original, precipitating error message either not to be overwritten by some operation performed between `PyErr_Fetch()` and `PyErr_Restore()`, or—as in this case—allow the original error message to be altered / extended.

Looking at this code, note that `newvalue` takes the return value of `PyUnicode_FromFormat()`/`PyString_FromFormat()` and previously `newvalue` was just unconditionally passed to `PyErr_Restore()`. In the case of OOM, `PyUnicode_FromFormat()`/`PyString_FromFormat()` \(which creates a new string object, and passes ownership of a reference to this string to the caller when an object is created\) can return `NULL` and fail to create a new string. While `PyErr_Restore()` handles being passed a `NULL` pointer for its `pvalue` argument, and so would not immediately crash as a result of this, we can note two consequences:

1. The Python error message has been corrupted, with `NULL` being substituted for `value`.
2. A reference to the string `value` is now leaked.

The loss of debug fidelity from the first issue is unfortunate, though perhaps could be argued as not compelling enough to necessitate a change—especially since it will only occur in OOM scenarios \(though I do think it's worth advocating for the ability to debug well!\).

However, I noted that we can avoid this loss of fidelity since we still have the original, referenced string in `value` and if in the case that no new string is created in `newvalue`, we instead restore `value` as the error message, then the problem of the leaked reference in `value` is also fixed as well. Given this analysis, I therefore thought it was prudent to address this issue as I was doing a sweep of the python object reference counting behaviour to look for other potential issues.

\(There are other OOM issues in the generated Python code which I haven't addressed in this PR and which would require more extensive changes. My intent was to consider those separately, and just fix any "easy" or "obvious" issues I found in this sweep.\)